pallets/manta-pay/src/lib.rs

```diff
@@ -480,19 +465,140 @@ pub mod pallet {
 	where
 		T: Config,
 	{
-		/// Returns the ledger implementation for this pallet.
+		/// Maximum Number of Updates per Shard
+		const PULL_MAX_PER_SHARD_UPDATE_SIZE: usize = 128;
+
+		/// Maximum Size of Sender Data Update
+		const PULL_MAX_SENDER_UPDATE_SIZE: usize = 1024;
+
+		/// Pulls receiver data from the ledger starting at the `receiver_index`.
 		#[inline]
-		fn ledger() -> Ledger<T> {
-			Ledger(PhantomData)
+		fn pull_receivers(receiver_index: &mut [usize; 256]) -> (bool, ReceiverChunk) {
+			let mut more_receivers = false;
+			let mut receivers = Vec::new();
+			for (i, index) in receiver_index.iter_mut().enumerate() {
+				more_receivers |= Self::pull_receivers_for_shard(i as u8, index, &mut receivers);
+			}
+			(more_receivers, receivers)
+		}
+
+		/// Pulls receiver data from the shard at `shard_index` starting at the `receiver_index`,
+		/// pushing the results back to `receivers`.
+		#[inline]
+		fn pull_receivers_for_shard(
+			shard_index: u8,
+			receiver_index: &mut usize,
+			receivers: &mut ReceiverChunk,
+		) -> bool {
+			let mut iter = if *receiver_index == 0 {
+				Shards::<T>::iter_prefix(shard_index)
+			} else {
+				let raw_key = Shards::<T>::hashed_key_for(shard_index, *receiver_index as u64 - 1);
+				Shards::<T>::iter_prefix_from(shard_index, raw_key)
+			};
+			for _ in 0..Self::PULL_MAX_PER_SHARD_UPDATE_SIZE {
+				match iter.next() {
+					Some((_, (utxo, encrypted_note))) => {
+						*receiver_index += 1;
+						receivers.push((utxo, encrypted_note));
+					}
+					_ => return false,
+				}
+			}
+			iter.next().is_some()
+		}
+
+		/// Pulls sender data from the ledger starting at the `sender_index`.
+		#[inline]
+		fn pull_senders(sender_index: &mut usize) -> (bool, SenderChunk) {
+			let mut senders = Vec::new();
+			let mut iter = InsertionOrderedVoidNumbers::<T>::iter().skip(*sender_index);
+			for _ in 0..Self::PULL_MAX_SENDER_UPDATE_SIZE {
+				match iter.next() {
+					Some((_, sender)) => {
+						*sender_index += 1;
+						senders.push(sender);
+					}
+					_ => return (false, senders),
+				}
+			}
+			(iter.next().is_some(), senders)
 		}
 
-		/// The account ID of AssetManager
+		/// Returns the update required to be synchronized with the ledger starting from
+		/// `checkpoint`.
+		#[inline]
+		pub fn pull(mut checkpoint: Checkpoint) -> PullResponse {
```

Comment:
I had to look through the code to understand what pull does, it's not obvious from the signature.
Can we rename it to something more concrete like `request_ledger_diff_from` or `fetch_ledger_updates_since`